### PR TITLE
[WIP] Add location and error reporting to macros PR

### PR DESCRIPTION
**Pull request by cd1m0**
In general this PR adds proper source location for errors in macros, proper locations for macro-derived annotations in instrumentation metadata, and some general cleanup and refactoring for how we handle locations in scribble. Specifically:

1. Change the internal `Location` type (and `Range` type) to directly reference the original file. Also add new `SourceFile` classes that can be used to this referecning. Move location/source classes and helper functions under `utils/locaiton.ts` and `utils/source.ts`

2. Change offsets, lines, columns inside `SNode`'s `src` locations to be relative to the begining of the file instead of the start of the annotation. This is now done at AST construction time, so we don't have to do conversions later on.

3. Change `SNode` locations to optionally be a pair of locations instead of a single location. If an`SNode`'s `src` field is a pair of locations, then that node must be a part of an annotation that was instantiated from a macro definition. The *first* location in the pair is the location of the annotation inside the macro file. The *second* location in the pair is the location of the macro call which instantiates the macro in the solidity file.

4. Tweak error printing in the CLI to be more readable.

5. Add Yaml schema checking and graceful errors when we detect a malformed yaml file.

Note: This PR is mostly finalized. There are only 2 things I may add tomorrow:

- unit tests for the yaml schema checking
- minor cleanup of 1-2 redundant fields in the various Error classes

### CODE REVIEW

File: src/util/yaml.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * 2. The top-level yaml node should have at least 1 entry "a" which is a scalar string
+ * 3. Any other entries in the top-level yaml node must be yaml maps too
+ * 4. Objects for other entries in the top-level yaml node must have two keys - "b" and "c"
+ * 5. Values for the "b" key must be scalar numbers
+ * 6. Values for the "c" key must be sequences, containing maps with a single key "d" which is a scalar string.
+ * @param node
+ * @param schema
+ * @param file
+ */
+export function checkYamlSchema(
```
Comment (blitz-1306) [inline src/util/yaml.ts:52]: Maybe we should consider dependency to do this for us.

File: src/util/yaml.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * Specifies that:
+ * 1. The top-level yaml node should be a map
+ * 2. The top-level yaml node should have at least 1 entry "a" which is a scalar string
+ * 3. Any other entries in the top-level yaml node must be yaml maps too
+ * 4. Objects for other entries in the top-level yaml node must have two keys - "b" and "c"
+ * 5. Values for the "b" key must be scalar numbers
+ * 6. Values for the "c" key must be sequences, containing maps with a single key "d" which is a scalar string.
+ * @param node
+ * @param schema
+ * @param file
```
Comment (blitz-1306) [inline src/util/yaml.ts:50]: Params are not described, while self explaindatory. Maybe remove them?

**Review by blitz-1306 — COMMENTED** [review]
Good work. Few clarifications and ideas. For now, I do not see blockers.

File: src/instrumenter/annotations.ts [hunk 3]
```diff
@@ -504,14 +359,14 @@ function findAnnotations(
...
-        text: raw.extractSourceFragment(source),
+        text: raw.extractSourceFragment(source.contents),
         loc: [sourceInfo.offset, sourceInfo.length]
     };
 
     const result: AnnotationMetaData[] = [];
 
     const rx =
-        /\s*(\*|\/\/\/)\s*#?(if_succeeds|if_updated|if_assigned|invariant|assert|try|require|((macro|define)\s*[a-zA-Z0-9_]*\s*\([^)]*\)))/g;
+        /\s*(\*|\/\/\/)\s*#?(if_succeeds|if_updated|if_assigned|invariant|assert|try|require|define\s*[a-zA-Z0-9_]*\s*\([^)]*\)|macro\s*[a-zA-Z0-9_]*\s*\([^)]*\))/g;
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/annotations.ts:369]: I must have mismereged during the re-base. Will fix this and check on the :msg section issue

File: src/instrumenter/type_string.ts [hunk 4]
```diff
@@ -35,15 +36,15 @@ export function makeTypeString(typeName: TypeName, loc: DataLocation): string {
     }
 
     if (typeName instanceof ArrayTypeName) {
-        assert(loc !== undefined, `${print(typeName)} requires location`);
+        assert(loc !== undefined, `{0} requires location`, typeName);
         const baseString = makeTypeString(typeName.vBaseType, loc);
         return `${baseString}[${
-            typeName.vLength !== undefined ? print(typeName.vLength) : ""
+            typeName.vLength !== undefined ? (typeName.vLength as Literal).value : ""
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/type_string.ts:42]: Nope thats a good catch! Will revert

File: src/macros/utils.ts [hunk 5]
```diff
@@ -24,37 +13,85 @@ export interface MacroVariable {
...
 
-    for (const [name, macro] of Object.entries(schema)) {
         const variables = new Map<string, MacroVariable>();
+        const varItems = macroVars === null ? [] : macroVars.items;
 
-        for (const [name, originalType] of Object.entries(macro.variables)) {
+        for (const variableItem of varItems) {
+            const name = variableItem.key.value as string;
+            const originalType = (variableItem.value as Scalar).value as string;
             const type: TypeNode = typeStringParse(originalType, {});
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/macros/utils.ts:65]: I think in the future we may type-check at variable types and function signature types at macro instantiation to provide better error messages. So lets leave this in for now

File: src/util/yaml.ts (hunk 1 above)
Comment (cd1m0) [reply src/util/yaml.ts:52]: I agree with you 100%. Maybe https://www.npmjs.com/package/yaml-schema-validator ?

Comment (cd1m0) [reply src/util/yaml.ts:52]: One nit: Our solution gives line/col information for schema mismatch errors. Preferrably any solution we find should do the same?

Comment (blitz-1306) [reply src/util/yaml.ts:52]: Lets postpone this to later stages. Packages that I look (including the one you already mentioned) are based on `js-yaml` and not `yaml`. So Scribble would install two YAML processing engines. We can pick proper package later.

File: src/instrumenter/annotations.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/annotations.ts]: Reverted the change in 8ae6a66. Still need to check if {:msg } works

File: src/instrumenter/type_string.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/type_string.ts]: Fixed in 8ae6a66.

File: src/spec-lang/expr_grammar.pegjs [hunk 6]
```diff
@@ -15,6 +15,11 @@ Annotation =
         / Macro
     )
     .* {
+        // Dummy statements to shut up unused variable warnings in tsc
+        /// @ts-ignore
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/expr_grammar.pegjs]: Removed in 8ae6a66.

File: src/util/yaml.ts (hunk 2 above)
Comment (cd1m0) [reply src/util/yaml.ts]: Fixed in 8ae6a66.

File: src/util/yaml.ts (hunk 1 above)
Comment (cd1m0) [reply src/util/yaml.ts:49]: Filed https://github.com/ConsenSys/scribble/issues/129 to track this so we don't forget

File: src/instrumenter/annotations.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/instrumenter/annotations.ts]: Actually we can fix this when we make the '#' required for annotations. Then we can simplify the regex.

**Review by blitz-1306 — APPROVED** [review]
Looks good.

**Pull request merged**